src/main/scala/com/fulcrumgenomics/umi/CollectDuplexSeqMetrics.scala

```diff
@@ -0,0 +1,540 @@
+
+    val umi1s = new ArrayBuffer[String]
+    val umi2s = new ArrayBuffer[String]
+
+    ab.iterator.map(r => r[String](RX).split('-')).foreach { case Array(u1, u2) => umi1s += u1; umi2s += u2 }
+    ba.iterator.map(r => r[String](RX).split('-')).foreach { case Array(u1, u2) => umi1s += u2; umi2s += u1 }
+
+    Seq(umi1s, umi2s).foreach { umis =>
+      val consensus = callConsensus(umis)
+      if (consensus.contains('N')) {
+        val x = 0
+      }
+
+      val metric    = this.umiMetricsMap.getOrElseUpdate(consensus, UmiMetric(umi=consensus))
+      metric.raw_observations    += umis.size
+      metric.unique_observations += 1
+      metric.raw_observations_with_errors += umis.filterNot(_ == consensus).size
+    }
+  }
+
+  /** Calls a simple consensus sequences from a set of sequences all the same length. */
+  private def callConsensus(umis: Seq[String]) = {
+    require(umis.nonEmpty, "Can't call consensus on an empty set of UMIs!")
+    val buffer = new StringBuilder
+
+    forloop(from=0, until=umis.head.length) { i =>
+      this.consensusBuilder.reset()
+      // The actual values for pError and pTruth don't matter since they're constant, but they are ln(0.01) and ln(0.99)
+      umis.foreach(umi => this.consensusBuilder.add(umi.charAt(i).toByte, pError = -4.60517, pTruth = -0.01005034))
+      val (base, qual) = this.consensusBuilder.call()
+      buffer.append(base.toChar)
+    }
+
+    buffer.toString()
+  }
+
+  /** Generates the family size metrics from the current observations. */
+  def familySizeMetrics: Seq[FamilySizeMetric] = {
+    val map = mutable.Map[Int, FamilySizeMetric]()
+    val startStopCounter = this.startStopFamilyCounter(1.0)
+    val duplexCounter    = this.duplexFamilyCounter(1.0)
+
+    // Add information about families grouped by genomic location alone
+    startStopCounter.foreach { case (size, count) =>
+      map.getOrElseUpdate(size, FamilySizeMetric(family_size=size)).cs_count += count
+    }
+
+    // Add information about the families grouped by ss and ds families
+    duplexCounter.foreach { case (Pair(ab, ba), count) =>
+      map.getOrElseUpdate(ab, FamilySizeMetric(family_size=ab)).ss_count += count
+      if (ba > 0) map.getOrElseUpdate(ba, FamilySizeMetric(family_size=ba)).ss_count += count
+      map.getOrElseUpdate(ab+ba, FamilySizeMetric(family_size=ab+ba)).ds_count += count
+    }
+
+    // Get a sorted Seq of metrics
+    val metrics = map.values.toIndexedSeq.sortBy(_.family_size)
+
+    // Fill in the fractions and cumulative fractions
+    val csTotal = metrics.map(_.cs_count).sum.toDouble
+    val ssTotal = metrics.map(_.ss_count).sum.toDouble
+    val dsTotal = metrics.map(_.ds_count).sum.toDouble
+
+    var csInverseCumulativeFraction = 0.0
+    var ssInverseCumulativeFraction = 0.0
+    var dsInverseCumulativeFraction = 0.0
+    metrics.foreach { m =>
+      m.cs_fraction = m.cs_count / csTotal
+      m.ss_fraction = m.ss_count / ssTotal
+      m.ds_fraction = m.ds_count / dsTotal
```

Comment:
The totals can only be zero if there are no observations at all, in which case `map` and `metrics` will both be empty so we never get into this `foreach`